KAFKA-2697: client-side support for leave group

hachikuji opened the pull request.

hachikuji committed:
KAFKA-2697: client-side support for leave group

hachikuji commented:
@onurkaraman @guozhangwang Ping for review. I made LeaveGroup best effort and only logged any errors. I also added a close(timeout) method on KafkaConsumer which mirrors the same method for KafkaProducer to give more control to users over shutdown. I'm adding additional test cases at the moment.

hachikuji committed:
rename consumerId to memberId in LeaveGroupRequest

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1149,10 +1149,15 @@ public void resume(TopicPartition... partitions) {
+        close(Long.MAX_VALUE, TimeUnit.MILLISECONDS);
+    }
+
+    public void close(long timeout, TimeUnit unit) {
```
Comment: we may now throw TimeoutException.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1167,11 +1172,17 @@ public void wakeup() {
+        try {
+            coordinator.close(timeoutMs);
+        } catch (Exception e) {
+            log.error("Failed to close coordinator", e);
```
We should still respect swallowException in closing coordinator by setting `firstException.compareAndSet(null, e)`.

ewencp replied:
Alternatively, should we add a `ClientUtils.closeQuietly()` overload that accepts a timeout as well? If we're eventually going to make this method correctly respect the timeout, we're going to need to have all these close methods accept timeout. Maybe some are guaranteed to finish quickly, but some have `await` calls in them, e.g. the `Metrics` close method. Might want to introduce something like a `TimeoutCloseable` interface in addition to `Closeable` to support this generally.

hachikuji replied:
Yeah, that's not a bad idea.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -299,17 +299,28 @@ public void ensurePartitionAssignment() {
+
+        long now = time.milliseconds();
+        long deadline = now + timeoutMs;
+        if (deadline < 0) deadline = Long.MAX_VALUE;
```
Could this ever happen with timeoutMS >= and now = time.milliseconds()? I think overflow should not happen as well.

hachikuji replied:
Not sure I follow. If timeoutMs = Long.MAX, then now + timeout would overflow, right?

guozhangwang replied:
I was originally thinking if `close()` is called we will just send a leave group request and wait on the response forever, and relying on the request timeout to return if the coordinator is ever not responding, i.e. calling `client.poll(request)` instead of `client.poll(request, timeout = Long.MAX)`.

hachikuji replied:
That's also what I had in mind initially, but it seemed like only a little more effort to give the user full control over shutdown, which I thought would be appreciated. That said, I'm also ok with the simpler blocking approach since, as you point out, it is bounded by the request timeout.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -331,24 +342,37 @@ public void onFailure(RuntimeException e) {
 
-            Utils.sleep(retryBackoffMs);
+            now = time.milliseconds();
+            long backoffDurationMs = Math.min(retryBackoffMs, deadline - now);
```
`deadline - now` could be negative.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java:
```diff
@@ -164,7 +164,8 @@ public void poll(RequestFuture<?> future) {
         long now = time.milliseconds();
         long deadline = now + timeout;
-        while (!future.isDone() && now < deadline) {
+        if (deadline < 0) deadline = Long.MAX_VALUE;
```
Similar above, could this happen?

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java:
```diff
@@ -299,17 +299,28 @@ public void ensurePartitionAssignment() {
-            } catch (WakeupException e) {
-                // ignore wakeups while closing to ensure we have a chance to commit
-                continue;
+    public void close(long timeoutMs) {
```
Will this function implicitly override AbstractCoordinator.close(long)?

onurkaraman commented:
Were you also planning on adding `void close(long timeout, TimeUnit unit)` to the Consumer interface?

hachikuji commented:
@onurkaraman Yeah, if we accept that method here, we probably should.

ewencp commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -201,6 +204,8 @@ protected boolean needRejoin() {
      */
     public void resetGeneration() {
+        // leave the group if it is needed, but don't block
+        maybeLeaveGroup(0);
```
Am I misreading this or will this, combined with the `TimeoutException` from `maybeLeaveGroup` cause this to always throw a `TimeoutException`? Some tests seem to be throwing it, so I think this might be the case and not the behavior you want here.

hachikuji replied:
Yeah, you are right. I missed that initially.

ewencp commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java:
```diff
@@ -537,6 +538,7 @@ public void testResetGeneration() {
         coordinator.ensurePartitionAssignment();
 
         // now switch to manual assignment
+        client.prepareResponse(new LeaveGroupResponse(Errors.NONE.code()).toStruct());
```
I was surprised this was the only addition to the tests. It seems we're missing a `testClose()`? That is by far the much more common scenario (`resetGeneration` looks like it is only used as a result of unsubscribing from all topics).

hachikuji replied:
I've got a few more test cases that I'll add shortly.

hachikuji committed:
rework timeout logic and add some test cases

hachikuji commented:
@guozhangwang I reworked the timeout logic a little bit, but it's still pretty ugly. I also modified the structure of the poll loop in KafkaConsumer.poll to use a similar do/while pattern since I noticed that the current behavior would result in delays of up to 1ms even when the passed timeout is 0. I wonder if that was the cause of the delays that you and Yasu had seen before.

hachikuji committed:
rename resetGeneration to maybeLeaveGroup

guozhangwang commented:
@hachikuji Could you rebase on trunk again?

hachikuji commented:
@guozhangwang Done.

hachikuji committed:
revert to simple blocking close

guozhangwang commented:
@hachikuji Is the jenkins failure on "could not merge" related? Otherwise LGTM overall.

guozhangwang commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -526,6 +519,67 @@ protected void coordinatorDead() {
+
+            @Override
+            public void onFailure(RuntimeException e) {
+                log.debug("LeaveGroup request failed with error", e);
```
Maybe bette to make it as log.info now?

hachikuji replied:
Or warn perhaps?

guozhangwang replied:
I feel INFO level is fine as in many cases people will keep alerting on warn / error logs.

hachikuji replied:
Fixed

hachikuji commented:
@guozhangwang It looks like it tried to merge an older commit. I rebased twice this morning fairly close together, so perhaps I triggered a race condition with the jenkins job? Any way to have it run again?

ewencp commented:
@hachikuji Closing and reopening the PR should trigger a rebuild on Jenkins.

hachikuji committed:
change log level for leave group failure to info

hachikuji commented:
Current build failure seems to be unrelated.
